policy_agent/policy_agent.py

```diff
@@ -0,0 +1,643 @@
+        self.kind = "Endpoints"
+        self.name = json["metadata"]["name"]
+        self.namespace = json["metadata"]["namespace"]
+        self.subsets = json["subsets"]
+
+    def get_key(self):
+        return "%s/%s" % (self.namespace, self.name)
+
+    def generate_svc_profiles_pods(self):
+        """
+        Endpoints objects contain a list of subsets
+        containing pod and policy info
+        subsets: [
+            {
+                addresses: [
+                    ...pods...
+                ],
+                ports: [
+                    ...associated ports...
+                ]
+            },
+            {
+                addresses: [
+                    ...pods...
+                ],
+                ports: [
+                    ...associated ports...
+                ]
+            },
+        ]
+        :return: A generated dict of profile-pod associations
+        :rtype: a dict of profiles mapping to a list of assoc pods
+        """
+        def verify_args(ports_obj, try_arg):
+            try:
+                arg = ports_obj[try_arg]
+            except KeyError:
+                arg = None
+            return arg
+
+        self.association_list = dict()
+        for subset in self.subsets:
+
+            # Generate a list of new svc profiles per port spec
+            port_rules = list()
+            for port_spec in subset["ports"]:
+                dst_port = verify_args(port_spec, "port")
+                protocol = verify_args(port_spec, "protocol").lower()
+                if protocol and dst_port:
+                    port_rules.append(Rule(action="allow",
+                                           dst_ports=[dst_port],
+                                           protocol=protocol))
+
+            # Create Profile for the subset
+            # TODO: better profile names
+            profile_name = "svc_%s_%s_port_%s_protocol_%s" % (
+                self.namespace, self.name, dst_port, protocol)
+            _log.info("Adding Profile %s to Store" % profile_name)
+
+            if not _datastore_client.profile_exists(profile_name):
+                _log.info("Creating Profile %s in Store" % profile_name)
+                profile_path = PROFILE_PATH % {"profile_id": profile_name}
+                _datastore_client.etcd_client.write(
+                    profile_path + "tags", '["%s"]' % profile_name)
+
+                # Determine rule set
+                default_allow = Rule(action="allow")
+                rules = Rules(id=profile_name,
+                              inbound_rules=port_rules,
+                              outbound_rules=[default_allow])
+
+                # Write rules to profile
+                _datastore_client.etcd_client.write(
+                    profile_path + "rules", rules.to_json())
+
+            else:
+                _log.warning("Profile %s already in Store" % profile_name)
+
+            # Generate list of pod names
+            pods = list()
```

Comment:
As a general note - Lists and dictionaries are usually declared like so:
```
somelist = []
dictionary = {}
```

I don't think there is anything wrong with the way you are doing it, but it sure looks weird to me. ;)